### PR TITLE
Test Host Extensibility Framework

### PR DESCRIPTION
**Pull request by mayankbansal018**
Added TestHost Manager Interface

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/vstest/507)

### CODE REVIEW

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 1]
```diff
@@ -61,6 +66,8 @@ public IDiscoveryRequest CreateDiscoveryRequest(DiscoveryCriteria discoveryCrite
             UpdateTestAdapterPaths(discoveryCriteria.RunSettings);
 
             var runconfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(discoveryCriteria.RunSettings);
+            //var testHostManager = this.testHostProviderManager.GetTestHostManagerByRunConfiguration(runconfiguration);
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: nit : this line of code can be removed.

**Comment by codito** [comment]
Reviewed 54 of 54 files at r1.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[src/Microsoft.TestPlatform.Client/TestPlatform.cs, line 32 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYbsRUvs9hjtgZxRI-:-KdYbsRVJMZN0ixFptBz:b35zny8) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Client/TestPlatform.cs#L32)):*
> ```Smalltalk
>         public TestPlatform() : this(new TestEngine())
>         {
>             this.testHostProviderManager = TestHostProviderManager.Instance;
> ```

Please inject this via ctor. I am not sure how we are going to unit test a static usage.

---

*[src/Microsoft.TestPlatform.Client/TestPlatform.cs, line 69 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYbAOxokX6WbTYCvyD-r1-69:-KdYbkLH6KGFwNETRo_M:bjbaex5) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Client/TestPlatform.cs#L69)):*
<details><summary><i>Previously, harshjain2 (Harsh Jain) wrote…</i></summary><blockquote>

nit : this line of code can be removed.
</blockquote></details>

Mark it with TODO if this is just the framework and you're staging the actual usage to later PR.

---

*[src/Microsoft.TestPlatform.Client/TestPlatform.cs, line 70 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYb_1VydJkDfnuS8E-:-KdYb_1VydJkDfnuS8E0:bqgrwp0) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Client/TestPlatform.cs#L70)):*
> ```Smalltalk
>             var runconfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(discoveryCriteria.RunSettings);
>             //var testHostManager = this.testHostProviderManager.GetTestHostManagerByRunConfiguration(runconfiguration);
>             
> ```

nit: remove trailing space

---

*[src/Microsoft.TestPlatform.Client/TestPlatform.cs, line 72 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYbd7aTm1PenY7O-99:-KdYbd7aTm1PenY7O-9A:bqgrwp0) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Client/TestPlatform.cs#L72)):*
> ```Smalltalk
>             
>             var testHostManager = this.TestEngine.GetDefaultTestHostManager(runconfiguration);
>             
> ```

nit: remove trailing space

---

*[src/Microsoft.TestPlatform.Common/Friends.cs, line 12 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYhtTxaYXegO81iVjh:-KdYhtTxaYXegO81iVji:b6syfxw) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Friends.cs#L12)):*
> ```Smalltalk
> [assembly: InternalsVisibleTo("Microsoft.TestPlatform.CommunicationUtilities, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
> [assembly: InternalsVisibleTo("Microsoft.VisualStudio.TestPlatform.Common, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
> [assembly: InternalsVisibleTo("Microsoft.VisualStudio.TestPlatform.Client, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
> ```

Not sure which type requires this. Why not make the type public?

---

*[src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs, line 68 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYcDFcxsePaMdvqU2p:-KdYcDFdj0A76dOA6NpY:b-1uje5m) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs#L68)):*
> ```Smalltalk
>                 TestSettingsProviders = new Dictionary<string, TestSettingsProviderPluginInformation>(StringComparer.OrdinalIgnoreCase),
>                 TestLoggers = new Dictionary<string, TestLoggerPluginInformation>(StringComparer.OrdinalIgnoreCase),
>                 TestHosts = new Dictionary<string, TestHostPluginInformation>(StringComparer.OrdinalIgnoreCase)
> ```

Is `TestHost` the concept or is it `RuntimeProvider`? My understanding is that it is later since a RuntimeProvider will provide ability to (a) deploy app, (b) instantiate the testhost on target, (c) ack that communication is established

This should probably be `TestRuntimePluginInformation`.

---

*[src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs, line 246 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYi6F4V69-DcRgl86n:-KdYi6F4V69-DcRgl86o:b-4l5vrx) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs#L246)):*
> ```Smalltalk
>                         this.GetTestExtensionFromType<TestLoggerPluginInformation>(type, typeof(ITestLogger), testLoggers);
>                         this.GetTestExtensionFromType<TestSettingsProviderPluginInformation>(type, typeof(ISettingsProvider), testSettingsProviders);
>                         this.GetTestExtensionFromType<TestHostPluginInformation>(type, typeof(ITestHostProvider), testHosts);
> ```

Do we need tests for this?

---

*[src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestHostPluginInformation.cs, line 11 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYcsuPW6S9sSP1c1DL:-KdYcsuPW6S9sSP1c1DM:bytfxpc) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestHostPluginInformation.cs#L11)):*
> ```Smalltalk
>     using System.Reflection;
> 
>     using ObjectModel;
> ```

Fully qualified namespace please.

---

*[src/Microsoft.TestPlatform.Common/Hosting/TestHostExtensionManager.cs, line 4 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYeDw7-zyd23P79r2U:-KdYeDw8BL_brGDTCIx7:binxh54) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Hosting/TestHostExtensionManager.cs#L4)):*
> ```Smalltalk
> // Licensed under the MIT license. See LICENSE file in the project root for full license information.
> 
> namespace Microsoft.VisualStudio.TestPlatform.Common.Hosting
> ```

We're using different namespaces for same concept? TP.Common.Hosting and OM.Host.

---

*[src/Microsoft.TestPlatform.Common/Hosting/TestHostMetadata.cs, line 14 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYeWwgBobC7IyB_5sP:-KdYeWwgBobC7IyB_5sQ:bqvva8r) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Hosting/TestHostMetadata.cs#L14)):*
> ```Smalltalk
>     {
>         /// <summary>
>         /// Constructor for TestHostMetadata
> ```

Resharper generates a beautiful stylecop compliant summary header. Suggest to use it.

---

*[src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs, line 33 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYeno7DbLSlxW4zz0W:-KdYeno7DbLSlxW4zz0X:b-vxbi52) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs#L33)):*
> ```Smalltalk
> 
>         private TestHostExtensionManager testHostExtensionManager;
>         
> ```

nit: remove trailing space.

---

*[src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs, line 56 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYetkxwoWQH9QJunJa:-KdYetkxwoWQH9QJunJb:b-lwbvu4) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs#L56)):*
> ```Smalltalk
>                 if (testHostManager == null)
>                 {
>                     lock (Synclock)
> ```

What is the scenario that this will be called in multiple threads?

---

*[src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs, line 60 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYfFaHuPUJ5Zv7HQIg:-KdYfFaHuPUJ5Zv7HQIh:b-olwjq8) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs#L60)):*
> ```Smalltalk
>                         if (testHostManager == null)
>                         {
>                             testHostManager = new TestHostProviderManager(TestSessionMessageLogger.Instance);
> ```

Should it use `this.messageLogger`?

---

*[src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs, line 86 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYfMUdHuKkPeHZhGC6:-KdYfMUegck0nhvOR7U8:b6zzob8) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs#L86)):*
> ```Smalltalk
>         }
> 
>         public ITestHostProvider GetTestHostManagerByRunConfiguration(RunConfiguration runConfiguarion)
> ```

nit: spelling

---

*[src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs, line 104 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYf3DFBr0XWqyIXSeD:-KdYf3DFBr0XWqyIXSeE:bv88oyx) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Hosting/TestHostProviderManager.cs#L104)):*
> ```Smalltalk
>             // Use SupressFinalize in case a subclass
>             // of this type implements a finalizer.
>             GC.SuppressFinalize(this);
> ```

Is there a finalizer? Please use the `disposing` and dispose overload pattern.

---

*[src/Microsoft.TestPlatform.Common/Interfaces/ITestHostCapabilities.cs, line 6 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYfTlCHoe5A-IehnlT:-KdYfTlCHoe5A-IehnlU:b-g1is8e) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Interfaces/ITestHostCapabilities.cs#L6)):*
> ```Smalltalk
> namespace Microsoft.VisualStudio.TestPlatform.Common.Interfaces
> {
>     using System.Diagnostics.CodeAnalysis;
> ```

Why is this namespace required?

---

*[src/Microsoft.TestPlatform.Common/Interfaces/ITestHostCapabilities.cs, line 13 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYfYROpEi_nA9D56rt:-KdYfYROpEi_nA9D56ru:bfgua17) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.Common/Interfaces/ITestHostCapabilities.cs#L13)):*
> ```Smalltalk
>     public interface ITestHostCapabilities : ITestExtensionCapabilities
>     {
>         /// specifies the friendly name corresponding to the TestHost.
> ```

Please use <summary> header.

---

*[src/Microsoft.TestPlatform.ObjectModel/Host/ITestHostManager.cs, line 23 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYfxFIxkwgkH9csfeU:-KdYfxFJmua1wtlxK2mI:b-dmtdya) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/src/Microsoft.TestPlatform.ObjectModel/Host/ITestHostManager.cs#L23)):*
> ```Smalltalk
> 
>         /// <summary>
>         /// Gets a value indicating whether the test host is specific to a test source. If yes, each test source
> ```

Comment is not clear.

---

*[test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostExtensionManagerTests.cs, line 18 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYg7aTufNgPzEcrlri:-KdYg7aTufNgPzEcrlrj:b-ajdacb) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostExtensionManagerTests.cs#L18)):*
> ```Smalltalk
>     {
>         [TestInitialize]
>         public void Initialize()
> ```

Use constructor instead of TestInitialize.

---

*[test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostExtensionManagerTests.cs, line 29 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYgB9S9zdHGisqeoqN:-KdYgB9S9zdHGisqeoqO:bf5cqla) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostExtensionManagerTests.cs#L29)):*
> ```Smalltalk
>                 var testLoggerExtensionManager = TestHostExtensionManager.Create(null);
>             });
> ```

nit: extra line

---

*[test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostExtensionManagerTests.cs, line 37 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYgunN1YwYNIpNbr1k:-KdYgunN1YwYNIpNbr1l:bl6xdle) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostExtensionManagerTests.cs#L37)):*
> ```Smalltalk
>             try
>             {
>                 var testLoggerExtensionManager = TestHostExtensionManager.Create(TestSessionMessageLogger.Instance);
> ```

No tests should use static instances. Please create `Testable` implementations and just test that.

---

*[test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostProviderManagerTests.cs, line 21 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYh6yDG_SVXHqquzql:-KdYh6yEMQKltnyheWm_:bhjr6sm) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostProviderManagerTests.cs#L21)):*
> ```Smalltalk
>     using TestPlatform.Common.UnitTests.ExtensionFramework;
>     using TestPlatform.Common.UnitTests.Utilities;
>     using ObjectModel = Microsoft.VisualStudio.TestPlatform.ObjectModel;
> ```

Why is this required? Are there conflicting OM namespaces?

---

*[test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostProviderManagerTests.cs, line 39 at r1](https://reviewable.io:443/reviews/microsoft/vstest/507#-KdYhF5aAcm1tEJUNhdg:-KdYhF5aAcm1tEJUNhdh:b-rgxtqo) ([raw file](https://github.com/microsoft/vstest/blob/bb3fc2ab5070026cbe5067ee9137f389cff7f6dc/test/Microsoft.TestPlatform.Common.UnitTests/Hosting/TestHostProviderManagerTests.cs#L39)):*
> ```Smalltalk
>         {
>             // Dispose the logger manager multiple times and verify that no exception is thrown.
>             var manager = TestHostProviderManager.Instance;
> ```

No tests should use static instances. Please create `Testable` implementation for testing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/vstest/507)*

**Review by codito — CHANGES_REQUESTED** [review]
Suggested few changes.

**Comment by codito** [comment]
@dotnet-bot test this please

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs [hunk 2]
```diff
@@ -63,7 +64,8 @@ internal class TestPluginDiscoverer
                 TestDiscoverers = new Dictionary<string, TestDiscovererPluginInformation>(StringComparer.OrdinalIgnoreCase),
                 TestExecutors = new Dictionary<string, TestExecutorPluginInformation>(StringComparer.OrdinalIgnoreCase),
                 TestSettingsProviders = new Dictionary<string, TestSettingsProviderPluginInformation>(StringComparer.OrdinalIgnoreCase),
-                TestLoggers = new Dictionary<string, TestLoggerPluginInformation>(StringComparer.OrdinalIgnoreCase)
+                TestLoggers = new Dictionary<string, TestLoggerPluginInformation>(StringComparer.OrdinalIgnoreCase),
+                TestHosts = new Dictionary<string, TestHostPluginInformation>(StringComparer.OrdinalIgnoreCase)
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginDiscoverer.cs]: We plan to rename TestHostPluginInformation to TestRuntimePluginInformation?

File: src/Microsoft.TestPlatform.Common/Hosting/TestRunTimeExtensionManager.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    using Microsoft.VisualStudio.TestPlatform.Common.Interfaces;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Host;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Manages loading and provides access to testhost extensions implementing the
+    /// ITestRunTimeProvider interface.
+    /// </summary>
+    internal class TestRunTimeExtensionManager : TestExtensionManager<ITestRunTimeProvider, ITestRunTimeCapabilities>
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Hosting/TestRunTimeExtensionManager.cs]: `TestRuntimeExtensionManager` since runtime is one word.

Comment (codito) [inline src/Microsoft.TestPlatform.Common/Hosting/TestRunTimeExtensionManager.cs]: Please update other usages of RunTime.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Hosting/DefaultTestHostManager.cs [hunk 4]
```diff
@@ -77,35 +119,49 @@ internal DefaultTestHostManager(Architecture architecture, Framework framework,
...
+        public async Task<int> LaunchTestHost(TestProcessStartInfo testHostStartInfo)
         {
-            this.DeregisterForExitNotification();
-            EqtTrace.Verbose("Launching default test Host Process {0} with arguments {1}", testHostStartInfo.FileName, testHostStartInfo.Arguments);
+            return await Task.Run(() => LaunchHost(testHostStartInfo), this.GetCancellationTokenSource().Token);
+        }
 
-            if (this.customTestHostLauncher == null)
+        /// <inheritdoc/>
+        public CancellationTokenSource GetCancellationTokenSource()
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Hosting/DefaultTestHostManager.cs]: This should be private.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Hosting/DotnetTestHostManager.cs [hunk 5]
```diff
@@ -41,11 +44,53 @@ public class DotnetTestHostManager : ITestHostManager
...
+        public event EventHandler<HostProviderEventArgs> HostLaunched;
+        public event EventHandler<HostProviderEventArgs> HostError;
+
+        protected int ErrorLength { get; set; } = 1000;
+        protected int TimeOut { get; set; } = 10000;
+
+        /// <summary>
+        /// Callback on process exit
+        /// </summary>
+        private Action<Process, string> ErrorReceivedCallback => ((process, data) =>
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Hosting/DotnetTestHostManager.cs]: DRY violation. Can this be extracted out?

**Review by codito — APPROVED** [review]
Other suggestions we plan to take in next PR:
- Remove Host namespace
- Add HostExited event to ITestRuntimeProvider

**Pull request merged**